File: scripts/trace_ut_computation.mjs
#!/usr/bin/env node
/**
 * Stage-wise trace for U_t computation in JavaScript.
 *
 * This script traces every step of U_t computation to identify divergence
 * between Python and JS implementations.
 *
 * Usage:
 *   node scripts/trace_ut_computation.mjs
 */

import { createHash } from 'crypto';
import { readFileSync, writeFileSync, mkdirSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = join(__dirname, '..');
const EXAMPLES_PATH = join(REPO_ROOT, 'site', 'v0.2.6', 'evidence-pack', 'examples.json');

// Domain separation constants (must match Python)
const DOMAIN_REASONING_LEAF = Buffer.concat([
    Buffer.from([0xA0]),
    Buffer.from('reasoning-leaf', 'utf8')
]);
const DOMAIN_UI_LEAF = Buffer.concat([
    Buffer.from([0xA1]),
    Buffer.from('ui-leaf', 'utf8')
]);
const DOMAIN_LEAF = Buffer.from([0x00]);
const DOMAIN_NODE = Buffer.from([0x01]);

// RFC 8785 JSON canonicalization
function can(o) {
    if (o === null) return 'null';
    if (typeof o === 'boolean') return o ? 'true' : 'false';
    if (typeof o === 'number') return Object.is(o, -0) ? '0' : String(o);
    if (typeof o === 'string') {
        let r = '"';
        for (let i = 0; i < o.length; i++) {
            const c = o.charCodeAt(i);
            if (c === 8) r += '\\b';
            else if (c === 9) r += '\\t';
            else if (c === 10) r += '\\n';
            else if (c === 12) r += '\\f';
            else if (c === 13) r += '\\r';
            else if (c === 34) r += '\\"';
            else if (c === 92) r += '\\\\';
            else if (c < 32) r += '\\u' + c.toString(16).padStart(4, '0');
            else r += o[i];
        }
        return r + '"';
    }
    if (Array.isArray(o)) return '[' + o.map(can).join(',') + ']';
    if (typeof o === 'object') {
        const k = Object.keys(o).sort();
        return '{' + k.map(x => can(x) + ':' + can(o[x])).join(',') + '}';
    }
    throw Error('bad');
}

// SHA256 with domain prefix (returns hex string)
function shaD(data, domain) {
    const dataBytes = typeof data === 'string' ? Buffer.from(data, 'utf8') : data;
    const combined = Buffer.concat([domain, dataBytes]);
    return createHash('sha256').update(combined).digest('hex');
}

// SHA256 with domain prefix (returns Buffer)
function shaDBytes(data, domain) {
    const dataBytes = typeof data === 'string' ? Buffer.from(data, 'utf8') : data;
    const combined = Buffer.concat([domain, dataBytes]);
    return createHash('sha256').update(combined).digest();
}

function traceUtComputation(uvilEvents) {
    const trace = {
        input_events: uvilEvents,
        event_traces: [],
        merkle_trace: null,
        final_u_t: null
    };

    const leafHashes = [];

    console.log('='.repeat(60));
    console.log('STAGE 1: Hash each UI event with DOMAIN_UI_LEAF');
    console.log('='.repeat(60));

    for (let i = 0; i < uvilEvents.length; i++) {
        const event = uvilEvents[i];
        const eventTrace = { index: i, original: event };

        // Step 1: RFC 8785 canonicalization
        const canonical = can(event);
        eventTrace.canonical_json = canonical;
        console.log(`\n[Event ${i}] RFC 8785 canonical:`);
        console.log(`  ${canonical}`);

        // Step 2: UTF-8 encode
        const canonicalBytes = Buffer.from(canonical, 'utf8');
        eventTrace.canonical_bytes_hex = canonicalBytes.toString('hex');
        console.log(`  Bytes (UTF-8): ${canonicalBytes.toString('hex')}`);

        // Step 3: Domain prefix
        eventTrace.domain_prefix_hex = DOMAIN_UI_LEAF.toString('hex');
        console.log(`  Domain prefix: ${DOMAIN_UI_LEAF.toString('hex')} (${DOMAIN_UI_LEAF})`);

        // Step 4: Full payload to hash
        const fullPayload = Buffer.concat([DOMAIN_UI_LEAF, canonicalBytes]);
        eventTrace.full_payload_hex = fullPayload.toString('hex');
        console.log(`  Full payload: ${fullPayload.toString('hex').slice(0, 40)}...`);

        // Step 5: SHA256 hash
        const leafHash = createHash('sha256').update(fullPayload).digest('hex');
        eventTrace.leaf_hash = leafHash;
        console.log(`  Leaf hash: ${leafHash}`);

        // Verify matches our shaD function
        const verifyHash = shaD(canonical, DOMAIN_UI_LEAF);
        eventTrace.shaD_result = verifyHash;
        if (leafHash !== verifyHash) {
            console.log(`  ERROR: Mismatch! ${leafHash} != ${verifyHash}`);
        } else {
            console.log(`  (verified: matches shaD)`);
        }

        leafHashes.push(leafHash);
        trace.event_traces.push(eventTrace);
    }

    console.log('\n' + '='.repeat(60));
    console.log('STAGE 2: Merkle tree construction');
    console.log('='.repeat(60));

    const merkleTrace = traceMerkleRoot(leafHashes);
    trace.merkle_trace = merkleTrace;
    trace.final_u_t = merkleTrace.final_root;

    return trace;
}

function traceMerkleRoot(leafHashes) {
    const trace = {
        input_leaf_hashes: leafHashes,
        after_sort: [],
        after_encode: [],
        levels: [],
        final_root: null
    };

    console.log(`\nInput leaf hashes (${leafHashes.length} items):`);
    for (let i = 0; i < leafHashes.length; i++) {
        console.log(`  [${i}] ${leafHashes[i]}`);
    }

    // Step 1: Sort the hex string hashes
    const sorted = [...leafHashes].sort();
    console.log('\n--- After sorting (strings) ---');
    for (let i = 0; i < sorted.length; i++) {
        trace.after_sort.push(sorted[i]);
        console.log(`  [${i}] ${sorted[i]}`);
    }

    // Step 2: UTF-8 encode each sorted hash string
    console.log('\n--- After UTF-8 encode ---');
    const encodedLeaves = [];
    for (let i = 0; i < sorted.length; i++) {
        const encoded = Buffer.from(sorted[i], 'utf8');
        encodedLeaves.push(encoded);
        trace.after_encode.push(encoded.toString('hex'));
        console.log(`  [${i}] ${encoded.toString('hex')}`);
    }

    // Step 3: Hash each leaf with DOMAIN_LEAF
    console.log('\n--- Hash with DOMAIN_LEAF ---');
    let nodes = [];
    const level0 = [];
    for (let i = 0; i < encodedLeaves.length; i++) {
        const fullPayload = Buffer.concat([DOMAIN_LEAF, encodedLeaves[i]]);
        const nodeHash = createHash('sha256').update(fullPayload).digest();
        nodes.push(nodeHash);
        level0.push({
            payload_hex: fullPayload.toString('hex'),
            hash_hex: nodeHash.toString('hex')
        });
        console.log(`  [${i}] payload: ${fullPayload.toString('hex').slice(0, 40)}...`);
        console.log(`       hash: ${nodeHash.toString('hex')}`);
    }
    trace.levels.push(level0);

    // Step 4: Build tree
    let levelNum = 1;
    while (nodes.length > 1) {
        console.log(`\n--- Level ${levelNum} (${nodes.length} -> ${Math.ceil(nodes.length / 2)}) ---`);

        if (nodes.length % 2 === 1) {
            nodes.push(nodes[nodes.length - 1]);
            console.log(`  (duplicated last node for odd count)`);
        }

        const nextLevel = [];
        const levelTrace = [];
        for (let i = 0; i < nodes.length; i += 2) {
            const combined = Buffer.concat([nodes[i], nodes[i + 1]]);
            const fullPayload = Buffer.concat([DOMAIN_NODE, combined]);
            const nodeHash = createHash('sha256').update(fullPayload).digest();
            nextLevel.push(nodeHash);
            levelTrace.push({
                left_hash: nodes[i].toString('hex'),
                right_hash: nodes[i + 1].toString('hex'),
                result_hash: nodeHash.toString('hex')
            });
            console.log(`  [${Math.floor(i / 2)}] L: ${nodes[i].toString('hex').slice(0, 16)}... R: ${nodes[i + 1].toString('hex').slice(0, 16)}...`);
            console.log(`       Result: ${nodeHash.toString('hex')}`);
        }

        trace.levels.push(levelTrace);
        nodes = nextLevel;
        levelNum++;
    }

    const finalRoot = nodes[0].toString('hex');
    trace.final_root = finalRoot;
    console.log(`\nFinal U_t: ${finalRoot}`);

    return trace;
}

function main() {
    // Load test data
    const examples = JSON.parse(readFileSync(EXAMPLES_PATH, 'utf8'));
    const validPack = examples.examples.valid_boundary_demo.pack;

    const uvilEvents = validPack.uvil_events;
    const expectedUt = validPack.u_t;

    console.log('='.repeat(60));
    console.log('U_t COMPUTATION TRACE (JavaScript)');
    console.log('='.repeat(60));
    console.log(`Expected U_t: ${expectedUt}`);
    console.log();

    const trace = traceUtComputation(uvilEvents);

    console.log('\n' + '='.repeat(60));
    console.log('VERIFICATION');
    console.log('='.repeat(60));
    console.log(`Expected: ${expectedUt}`);
    console.log(`Computed: ${trace.final_u_t}`);
    if (trace.final_u_t === expectedUt) {
        console.log('STATUS: MATCH!');
    } else {
        console.log('STATUS: MISMATCH!');
    }

    // Save trace for comparison
    const tmpDir = join(REPO_ROOT, 'tmp');
    try { mkdirSync(tmpDir, { recursive: true }); } catch (e) {}
    const tracePath = join(tmpDir, 'ut_trace_js.json');
    writeFileSync(tracePath, JSON.stringify(trace, null, 2));
    console.log(`\nTrace saved to: ${tracePath}`);
}

main();

File: scripts/trace_ut_computation.py
#!/usr/bin/env python3
"""
Stage-wise trace for U_t computation.

This script traces every step of U_t computation to identify divergence
between Python and JS implementations.

Usage:
    uv run python scripts/trace_ut_computation.py
"""

import json
import hashlib
from pathlib import Path

# Import the actual functions used in Python
from substrate.crypto.core import rfc8785_canonicalize
from substrate.crypto.hashing import sha256_hex, sha256_bytes, DOMAIN_LEAF, DOMAIN_NODE
from attestation.dual_root import DOMAIN_UI_LEAF, hash_ui_leaf
from normalization.canon import normalize

REPO_ROOT = Path(__file__).parent.parent
EXAMPLES_PATH = REPO_ROOT / "site" / "v0.2.6" / "evidence-pack" / "examples.json"


def trace_ut_computation(uvil_events: list) -> dict:
    """Trace every step of U_t computation."""
    trace = {
        "input_events": uvil_events,
        "event_traces": [],
        "merkle_trace": None,
        "final_u_t": None,
    }

    leaf_hashes = []

    print("=" * 60)
    print("STAGE 1: Hash each UI event with DOMAIN_UI_LEAF")
    print("=" * 60)

    for i, event in enumerate(uvil_events):
        event_trace = {"index": i, "original": event}

        # Step 1: RFC 8785 canonicalization
        canonical = rfc8785_canonicalize(event)
        event_trace["canonical_json"] = canonical
        print(f"\n[Event {i}] RFC 8785 canonical:")
        print(f"  {canonical}")

        # Step 2: UTF-8 encode
        canonical_bytes = canonical.encode('utf-8')
        event_trace["canonical_bytes_hex"] = canonical_bytes.hex()
        print(f"  Bytes (UTF-8): {canonical_bytes.hex()}")

        # Step 3: Domain prefix
        domain_bytes = DOMAIN_UI_LEAF
        event_trace["domain_prefix_hex"] = domain_bytes.hex()
        print(f"  Domain prefix: {domain_bytes.hex()} ({domain_bytes!r})")

        # Step 4: Full payload to hash (domain + data)
        full_payload = domain_bytes + canonical_bytes
        event_trace["full_payload_hex"] = full_payload.hex()
        print(f"  Full payload: {full_payload.hex()[:40]}...")

        # Step 5: SHA256 hash
        leaf_hash = hashlib.sha256(full_payload).hexdigest()
        event_trace["leaf_hash"] = leaf_hash
        print(f"  Leaf hash: {leaf_hash}")

        # Verify matches our hash_ui_leaf function
        verify_hash = hash_ui_leaf(canonical)
        event_trace["hash_ui_leaf_result"] = verify_hash
        assert leaf_hash == verify_hash, f"Mismatch! {leaf_hash} != {verify_hash}"
        print(f"  (verified: matches hash_ui_leaf)")

        leaf_hashes.append(leaf_hash)
        trace["event_traces"].append(event_trace)

    print("\n" + "=" * 60)
    print("STAGE 2: Merkle tree construction")
    print("=" * 60)

    merkle_trace = trace_merkle_root(leaf_hashes)
    trace["merkle_trace"] = merkle_trace
    trace["final_u_t"] = merkle_trace["final_root"]

    return trace


def trace_merkle_root(leaf_hashes: list) -> dict:
    """Trace merkle_root computation step by step."""
    trace = {
        "input_leaf_hashes": leaf_hashes,
        "after_normalize": [],
        "after_encode": [],
        "sorted_leaves_hex": [],
        "levels": [],
        "final_root": None,
    }

    print(f"\nInput leaf hashes ({len(leaf_hashes)} items):")
    for i, lh in enumerate(leaf_hashes):
        print(f"  [{i}] {lh}")

    # Step 1: normalize() each leaf hash
    print("\n--- After normalize() ---")
    for i, lh in enumerate(leaf_hashes):
        normalized = normalize(lh)
        trace["after_normalize"].append(normalized)
        changed = "(CHANGED!)" if normalized != lh else "(unchanged)"
        print(f"  [{i}] {normalized} {changed}")

    # Step 2: UTF-8 encode
    print("\n--- After UTF-8 encode ---")
    encoded_leaves = []
    for i, norm_lh in enumerate(trace["after_normalize"]):
        encoded = norm_lh.encode('utf-8')
        encoded_leaves.append(encoded)
        trace["after_encode"].append(encoded.hex())
        print(f"  [{i}] {encoded.hex()}")

    # Step 3: Sort
    sorted_leaves = sorted(encoded_leaves)
    print("\n--- After sorting ---")
    for i, sl in enumerate(sorted_leaves):
        trace["sorted_leaves_hex"].append(sl.hex())
        print(f"  [{i}] {sl.hex()}")

    # Step 4: Hash each leaf with DOMAIN_LEAF
    print("\n--- Hash with DOMAIN_LEAF ---")
    nodes = []
    level_0 = []
    for i, leaf_bytes in enumerate(sorted_leaves):
        full_payload = DOMAIN_LEAF + leaf_bytes
        node_hash = hashlib.sha256(full_payload).digest()
        nodes.append(node_hash)
        level_0.append({
            "payload_hex": full_payload.hex(),
            "hash_hex": node_hash.hex()
        })
        print(f"  [{i}] payload: {full_payload.hex()[:40]}...")
        print(f"       hash: {node_hash.hex()}")
    trace["levels"].append(level_0)

    # Step 5: Build tree
    level_num = 1
    while len(nodes) > 1:
        print(f"\n--- Level {level_num} ({len(nodes)} -> {(len(nodes) + 1) // 2}) ---")

        if len(nodes) % 2 == 1:
            nodes.append(nodes[-1])
            print(f"  (duplicated last node for odd count)")

        next_level = []
        level_trace = []
        for i in range(0, len(nodes), 2):
            combined = nodes[i] + nodes[i + 1]
            full_payload = DOMAIN_NODE + combined
            node_hash = hashlib.sha256(full_payload).digest()
            next_level.append(node_hash)
            level_trace.append({
                "left_hash": nodes[i].hex(),
                "right_hash": nodes[i + 1].hex(),
                "result_hash": node_hash.hex()
            })
            print(f"  [{i//2}] L: {nodes[i].hex()[:16]}... R: {nodes[i+1].hex()[:16]}...")
            print(f"       Result: {node_hash.hex()}")

        trace["levels"].append(level_trace)
        nodes = next_level
        level_num += 1

    final_root = nodes[0].hex()
    trace["final_root"] = final_root
    print(f"\nFinal U_t: {final_root}")

    return trace


def main():
    # Load test data
    examples = json.loads(EXAMPLES_PATH.read_text(encoding="utf-8"))
    valid_pack = examples["examples"]["valid_boundary_demo"]["pack"]

    uvil_events = valid_pack["uvil_events"]
    expected_u_t = valid_pack["u_t"]

    print("=" * 60)
    print("U_t COMPUTATION TRACE")
    print("=" * 60)
    print(f"Expected U_t: {expected_u_t}")
    print()

    trace = trace_ut_computation(uvil_events)

    print("\n" + "=" * 60)
    print("VERIFICATION")
    print("=" * 60)
    print(f"Expected: {expected_u_t}")
    print(f"Computed: {trace['final_u_t']}")
    if trace['final_u_t'] == expected_u_t:
        print("STATUS: MATCH!")
    else:
        print("STATUS: MISMATCH!")

    # Save trace for comparison
    trace_path = REPO_ROOT / "tmp" / "ut_trace_python.json"
    trace_path.parent.mkdir(exist_ok=True)
    with open(trace_path, 'w') as f:
        json.dump(trace, f, indent=2)
    print(f"\nTrace saved to: {trace_path}")


if __name__ == "__main__":
    main()
